random_generator/utils/atl_functions.py

```diff
@@ -0,0 +1,112 @@
+from enum import IntEnum
+
+import numpy as np
+from utils.configuration import Config
+from utils.utility_functions import write_json
+
+
+class Enforce(IntEnum):
+    enforce = 0
+    despite = 1
+
+    def __str__(self):
+        return self.name
+
+
+class WhateverThisIsCalled(IntEnum):
+    next = 0
+    eventually = 1
+    invariant = 2
+    until = 3
+
+    def __str__(self):
+        return self.name
+
+
+def get_next(labels: list[int], next_state_list, states) -> (np.array, np.array):
+    next_state = np.array([])
+    labels_here = np.array([])
+    for state in states:
+        next_state = np.append(next_state, next_state_list[int(state)])
+        labels_here = np.append(labels_here, labels[int(state)])
+    return labels_here, next_state
+
+
+def early_termination(config: Config,
+                      labels: list[int],
+                      predecessors_list: dict):
+    initial_available_labels, initial_available_next_sates = get_next(labels, predecessors_list, [0])
+    non_initial_labels = [x for x in range(0, 12) if x not in initial_available_labels]
+
+    atl_queries = {}
+    for index in range(config.amount_of_atl_formulas):
+        for true_or_false in [True, False]:
+            for k in Enforce:
+                for o in [x for x in WhateverThisIsCalled
+                          if x not in [WhateverThisIsCalled.eventually, WhateverThisIsCalled.until]
+                          and (x != WhateverThisIsCalled.invariant or not true_or_false)]:
```

Comment:
Can you explain this condition?

Reply:
Yes, for it to be an early termination, we need to only use the temporaloperators that does not require us to search through the whole statespace.
Eg. eventually, and until, can be true at the end. And the invariant can only be used if it is false, else we would have to use the whole statespace.

I wrote a comment about it in the code as well, as this could easily be forgotten if it was just a github comment